packages/vite/src/node/utils.ts

```diff
@@ -663,3 +663,26 @@ export function parseRequest(id: string): Record<string, string> | null {
   }
   return Object.fromEntries(new URLSearchParams(search.slice(1)))
 }
+
+// TS rewrite of https://github.com/rxaviers/async-pool/blob/master/lib/es7.js
+// Credits: Rafael Xavier de Souza http://rafael.xavier.blog.br
+export async function asyncPool<Item, Result>({
+  concurrency,
+  items,
+  fn
+}: {
+  concurrency: number
+  items: Item[]
+  fn: (item: Item) => Promise<Result>
+}): Promise<Result[]> {
+  const promises: Promise<Result>[] = []
+  const pool = new Set<Promise<Result>>()
+  for (const item of items) {
+    const promise = fn(item)
+    promises.push(promise)
+    pool.add(promise)
+    promise.then(() => pool.delete(promise))
+    if (pool.size >= concurrency) await Promise.race(pool)
+  }
+  return Promise.all(promises)
+}
```

Comment:
really cool implementation.
whenever i need a pool of promises, i use `throat` simple lib https://www.npmjs.com/package/throat

```
const throat = require('throat')(2);

const resA = throat(async () => { /* async stuff... */ });
const resB = throat(async () => { /* async stuff... */ });

```

or 

```
const throat = require('throat');

const input = ['fileA.txt', 'fileB.txt', 'fileC.txt', 'fileD.txt'];
const data = Promise.all(
  input.map(throat(2, (fileName) => readFile(fileName)))
);

```

Reply:
Didn't know, the first version can be useful to manage a pool across multiples files!

I know the vite team avoid adding too much dependencies, this is why I added this implementation instead of adding a dependency.